File: src/main/java/me/zeroeightsix/kami/command/commands/BindCommand.java
package me.zeroeightsix.kami.command.commands;

import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.command.syntax.ChunkBuilder;
import me.zeroeightsix.kami.command.syntax.parsers.EnumParser;
import me.zeroeightsix.kami.command.syntax.parsers.ModuleParser;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.module.ModuleManager;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.setting.builder.SettingBuilder;
import me.zeroeightsix.kami.util.Wrapper;

import static me.zeroeightsix.kami.util.text.MessageSendHelper.*;

/**
 * Created by 086 on 12/11/2017.
 */
public class BindCommand extends Command {

    public static Setting<Boolean> modifiersEnabled = SettingBuilder.register(Settings.b("modifiersEnabled", false), "binds");

    public BindCommand() {
        super("bind", new ChunkBuilder()
                .append("[module]|modifiers|list", true, new ModuleParser())
                .append("state", false, new EnumParser(new String[]{"key", "on", "off"}))
                .build()
        );
        setDescription("Binds a module to a key, or allows you to change modifier options");
    }

    @Override
    public void call(String[] args) {
        if (args.length == 1) {
            sendChatMessage("Please specify a module.");
            return;
        }

        String module = args[0];
        String rkey = args[1];

        if (module.equalsIgnoreCase("list")) {
            Module[] modules = ModuleManager.getModules();

            sendChatMessage("You have the following binds: ");
            for (Module module1 : modules) {
                if (module1.getBind().getValue().toString().equals("None")) continue;
                sendRawChatMessage(module1.getBind().getValue().toString() + ": " + module1.getName().getValue());
            }
            return;
        } else if (module.equalsIgnoreCase("modifiers")) {
            if (rkey == null) {
                sendChatMessage("Expected: on or off");
                return;
            }

            if (rkey.equalsIgnoreCase("on")) {
                modifiersEnabled.setValue(true);
                sendChatMessage("Turned modifiers on.");
            } else if (rkey.equalsIgnoreCase("off")) {
                modifiersEnabled.setValue(false);
                sendChatMessage("Turned modifiers off.");
            } else {
                sendChatMessage("Expected: on or off");
            }
            return;
        }

        try {
            Module m = ModuleManager.getModule(module);
            if (rkey == null) {
                sendChatMessage(m.getBind().getValue().toString() + " is bound to &b" + m.getBindName());
                return;
            }
            int key = Wrapper.getKey(rkey);
            if (rkey.equalsIgnoreCase("none")) {
                key = -1;
            }
            if (key == 0) {
                sendErrorMessage("Unknown key '&7" + rkey + "&f'! Left alt is &7lmenu&f, left Control is &7lcontrol&f and ` is &7grave&f. You cannot bind the &7meta&f key.");
                return;
            }
            m.getBind().getValue().setKey(key);
            sendChatMessage("Bind for &b" + m.getName().getValue() + "&r set to &b" + rkey.toUpperCase());
        } catch (ModuleManager.ModuleNotFoundException x) {
            sendChatMessage("Unknown module '" + module + "'!");
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/command/commands/BindCommand.kt
package me.zeroeightsix.kami.command.commands

import me.zeroeightsix.kami.command.ClientCommand
import me.zeroeightsix.kami.module.ModuleManager
import me.zeroeightsix.kami.module.modules.client.CommandConfig
import me.zeroeightsix.kami.util.KeyboardUtils
import me.zeroeightsix.kami.util.text.MessageSendHelper
import me.zeroeightsix.kami.util.text.format
import me.zeroeightsix.kami.util.text.formatValue
import net.minecraft.util.text.TextFormatting

object BindCommand : ClientCommand(
    name = "bind",
    description = "Bind and unbind modules"
) {
    init {
        literal("list") {
            execute("List used module binds") {
                val modules = ModuleManager.getModules().filter { it.bind.value.key > 0 }.sortedBy { it.bind.toString() }

                MessageSendHelper.sendChatMessage("Used binds: ${formatValue(modules.size)}")
                modules.forEach {
                    MessageSendHelper.sendRawChatMessage("${formatValue(it.bind)} ${it.name}")
                }
            }
        }

        literal("reset", "unbind") {
            module("module") { moduleArg ->
                execute("Reset the bind of a module to nothing") {
                    val module = moduleArg.value
                    module.bind.resetValue()
                    MessageSendHelper.sendChatMessage("Reset bind for ${module.name}!")
                }
            }
        }

        literal("modifiers") {
            boolean("enabled") { modifiersArg ->
                execute("Disallow binds while holding a modifier") {
                    val modifiers = modifiersArg.value

                    CommandConfig.modifierEnabled.value = modifiers
                    MessageSendHelper.sendChatMessage(
                        "Modifiers ${if (modifiers) " ${TextFormatting.GREEN format "enabled"}" else " ${TextFormatting.RED format "disabled"}"}"
                    )
                }
            }
        }

        module("module") { moduleArg ->
            string("bind") { bindArg ->
                execute("Bind a module to a key") {
                    val module = moduleArg.value
                    val bind = bindArg.value

                    if (bind.equals("none", true)) {
                        module.bind.resetValue()
                        MessageSendHelper.sendChatMessage("Reset bind for ${module.name}!")
                        return@execute
                    }


                    val key = KeyboardUtils.getKey(bind)
                    if (key <= 0) {
                        KeyboardUtils.sendUnknownKeyError(bind)
                    } else {
                        module.bind.setValue(bind)
                        MessageSendHelper.sendChatMessage("Bind for ${module.name} set to ${formatValue(module.bind)}!")
                    }
                }
            }

            execute("Get the bind of a module") {
                val module = moduleArg.value
                MessageSendHelper.sendChatMessage("${module.name} is bound to ${formatValue(module.bind)}")
            }
        }
    }
}